controllers/operatorpolicy_controller.go

```diff
@@ -1198,22 +1216,96 @@ func (r *OperatorPolicyReconciler) considerResolutionFailed(
+			policy.Status.SubscriptionInterventionTime = nil
+		}
 
 		return mergedSub, nil, changed, nil
 	}
 
 	// A "constraints not satisfiable" message has nondeterministic clauses, and can be noisy with a list of versions.
 	// Just report a generic condition, which will prevent the OperatorPolicy status from constantly updating
 	// when the details in the Subscription status change.
-	changed := updateStatus(policy, subConstraintsNotSatisfiableCond, nonCompObj(foundSub, "ConstraintsNotSatisfiable"))
+	changed := updateStatus(policy, subConstraintsNotSatisfiableCond,
+		nonCompObj(mergedSub, "ConstraintsNotSatisfiable"))
+
+	if policy.Status.SubscriptionInterventionExpired() || policy.Status.SubscriptionInterventionTime == nil {
+		interventionTime := metav1.Time{Time: time.Now().Add(olmGracePeriod)}
+		policy.Status.SubscriptionInterventionTime = &interventionTime
+
+		opLog.V(1).Info("Detected ConstraintsNotSatisfiable, setting an intervention time",
+			"interventionTime", interventionTime, "subscription", mergedSub)
+
+		return mergedSub, nil, changed, nil
+	}
+
+	if policy.Status.SubscriptionInterventionWaiting() {
+		opLog.V(1).Info("Detected ConstraintsNotSatisfiable, giving OLM more time before possibly intervening",
+			"interventionTime", policy.Status.SubscriptionInterventionTime)
+
+		return mergedSub, nil, changed, nil
+	}
+
+	// Do the "intervention"
 
-	return mergedSub, nil, changed, nil
+	relatedCSVs := policy.Status.RelatedObjsOfKind("ClusterServiceVersion")
+	if len(relatedCSVs) != 1 {
+		opLog.V(1).Info("No related CSVs to use for an intervention")
+
+		return mergedSub, nil, changed, nil
+	}
+
+	csvIdx := 0
+	for idx := range relatedCSVs {
+		csvIdx = idx
+	}
+
+	csvName := relatedCSVs[csvIdx].Object.Metadata.Name
+
+	matches := unreferencedCSVRegex.FindStringSubmatch(subResFailed.Message)
+	if len(matches) < 2 {
+		opLog.V(1).Info("Subscription condition does not match pattern for an unreferenced CSV",
+			"subscriptionConditionMessage", subResFailed.Message)
+
+		return mergedSub, nil, changed, nil
+	}
+
+	if csvName != matches[1] {
+		opLog.V(1).Info("The CSV found in the Subscription condition does not match the CSV found by label",
+			"subscriptionConditionMessage", subResFailed.Message, "csvName", csvName, "csvInSub", matches[1])
+
+		return mergedSub, nil, changed, nil
+	}
+
+	if mergedSub.Status.LastUpdated.IsZero() {
+		mergedSub.Status.LastUpdated = metav1.Now()
+	}
+
+	mergedSub.Status.CurrentCSV = csvName
+
+	opLog.Info("Updating Subscription status to point to CSV", "csvName", csvName)
+
+	if err := r.TargetClient.Status().Update(ctx, mergedSub); err != nil {
+		return mergedSub, nil, changed,
+			fmt.Errorf("error updating the Subscription status to point to the CSV: %w", err)
+	}
+
+	mergedSub.SetGroupVersionKind(subscriptionGVK) // Update might strip this information
+
+	updateStatus(policy, updatedCond("Subscription"), updatedObj(mergedSub))
+
+	return mergedSub, nil, true, nil
```

Comment:
Should `policy.Status.SubscriptionInterventionTime` be set to empty after the update is successful and before returning?

Reply:
Originally I had this just apply the intervention once, then wait another 30 seconds again before trying again. What I found in testing was that about 5% of the time, OLM would then add a different condition, remove our update, and I think confuse itself... I still don't know what exactly was happening there, but the new implementation will now immediately re-apply the intervention in that case, as long as 10 seconds haven't passed. And it requires not clearing the timestamp until those 10 seconds are up.

Reply:
The weirdest thing looking back at it is that after it got into that weird situation of confusing itself, the controller would re-apply the intervention 30 seconds later, but OLM seemed to do the same weird thing *consistently* this time (not a 5% chance like I'd expect). It wasn't just a bad sync of timing, because I tried a different grace period and it would still happen.